daemon/graphdriver/driver.go

```diff
@@ -142,10 +142,40 @@ func GetDriver(name, home string, options []string) (Driver, error) {
 func New(root string, options []string) (driver Driver, err error) {
 	for _, name := range []string{os.Getenv("DOCKER_DRIVER"), DefaultDriver} {
 		if name != "" {
+			logrus.Infof("[graphdriver] trying provided driver %q", name) // so the logs show specified driver
 			return GetDriver(name, root, options)
 		}
 	}
 
+	// Guess for prior driver
+	priorDrivers := scanPriorDrivers(root)
+	for _, name := range priority {
+		if name == "vfs" {
+			// don't use vfs even if there is state present.
+			continue
+		}
+		for _, prior := range priorDrivers {
+			// of the state found from prior drivers, check in order of our priority
+			// which we would prefer
+			if prior == name {
+				driver, err = GetDriver(name, root, options)
+				if err != nil {
+					// unlike below, we will return error here, because there is prior
+					// state, and now it is no longer supported/prereq/compatible, so
+					// something changed and needs attention. Otherwise the daemon's
+					// images would just "disappear".
+					logrus.Errorf("[graphdriver] prior storage driver %q failed: %s", name, err)
+					return nil, err
+				}
+				if err := checkPriorDriver(name, root); err != nil {
```

Comment:
fair, but it is not a big hit to loop again, and this way we keep priority
list, and can exclude vfs, and checkfor additional drivers. I bet it's
cheaper than creating a map[string]string to compare against.

On Wed, Apr 8, 2015 at 2:42 PM, Jessie Frazelle <notifications@github.com>
wrote:

> In daemon/graphdriver/driver.go
> <https://github.com/docker/docker/pull/11999#discussion_r28000116>:
>
> > +			continue
> > +		}
> > +		for _, prior := range priorDrivers {
> > +			// of the state found from prior drivers, check in order of our priority
> > +			// which we would prefer
> > +			if prior == name {
> > +				driver, err = GetDriver(name, root, options)
> > +				if err != nil {
> > +					// unlike below, we will return error here, because there is prior
> > +					// state, and now it is no longer supported/prereq/compatible, so
> > +					// something changed and needs attention. Otherwise the daemon's
> > +					// images would just "disappear".
> > +					logrus.Errorf("[graphdriver] prior storage driver %q failed: %s", name, err)
> > +					return nil, err
> > +				}
> > +				if err := checkPriorDriver(name, root); err != nil {
>
> still kinda feels unnecessary that this function calls scanPriorDrivers
> again which we just did above
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11999/files#r28000116>.
>